### PR TITLE
Vector norm

### PR DESCRIPTION
**Pull request by wolfv**
Added Norm.

I had to modify to_sympy_matrix or rather matrix_data to also accept lists (ie. vectors, which are also matrices in sympy).
Not sure if this has unintended consequences. 

Cheers,

Wolf

### CODE REVIEW

**Comment by wolfv** [comment]
Hmm, didn't mean to include the lists.py changes :/

File: mathics/builtin/lists.py [hunk 1]
```diff
@@ -2012,3 +2012,113 @@ def get_call():
...
+        Accumulate values from list and return new list
+        </dd>
+        </dt>
+    </dl>
+    >> Accumulate[{1, 2, 3}]
+     = {1, 3, 6}
+    """
+
+    rules = {
+        'Accumulate[list_List]': 'FoldList[Plus, list]'
```
Comment (sn6uv) [inline mathics/builtin/lists.py]: According to the Mathematica documentation, "`Accumulate[list]` is effectively equivalent to `Rest[FoldList[Plus,0,list]]`."

I see two minor issues with this approach:

- `Accumulate` should work for any head (not just `List`). e.g.
```
In[1] := Accumulate[f[a, b, c, d]]
Out[1]= f[a, a + b, a + b + c, a + b + c + d]
```

- This fails on the empty list
```
In[3]:= Accumulate[{}]
Traceback (most recent call last):
  File "mathics/main.py", line 324, in <module>
    main()
  File "mathics/main.py", line 315, in main
    shell.evaluate(total_input)
  File "mathics/main.py", line 120, in evaluate
    out_callback=out_callback)
  File "/home/angus/prog/Mathics/mathics/core/evaluation.py", line 207, in __init__
    result = evaluate()
  File "/home/angus/prog/Mathics/mathics/core/evaluation.py", line 192, in evaluate
    result = query.evaluate(self)
  File "mathics/core/expression.py", line 776, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:22266)
    result = result.evaluate(evaluation)
  File "mathics/core/expression.py", line 776, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:22266)
    result = result.evaluate(evaluation)
  File "mathics/core/expression.py", line 726, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:21079)
    leaves[index] = leaves[index].evaluate(evaluation)
  File "mathics/core/expression.py", line 773, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:22179)
    result = rule.apply(new, evaluation, fully=False)
  File "mathics/core/rules.py", line 115, in mathics.core.rules.BaseRule.apply (mathics/core/rules.c:2677)
    self.pattern.match(
  File "mathics/core/pattern.py", line 224, in mathics.core.pattern.ExpressionPattern.match (mathics/core/pattern.c:6424)
    self.head.match(
  File "mathics/core/pattern.py", line 153, in mathics.core.pattern.AtomPattern.match (mathics/core/pattern.c:4758)
    yield_func(vars, None)
  File "mathics/core/pattern.py", line 216, in mathics.core.pattern.ExpressionPattern.match.yield_head (mathics/core/pattern.c:6024)
    self.get_pre_choices(
  File "mathics/core/pattern.py", line 341, in mathics.core.pattern.ExpressionPattern.get_pre_choices (mathics/core/pattern.c:8974)
    yield_func(vars)
  File "mathics/core/pattern.py", line 201, in mathics.core.pattern.ExpressionPattern.match.yield_choice (mathics/core/pattern.c:5833)
    self.match_leaf(
  File "mathics/core/pattern.py", line 506, in mathics.core.pattern.ExpressionPattern.match_leaf (mathics/core/pattern.c:12726)
    self.get_wrappings(
  File "mathics/core/pattern.py", line 362, in mathics.core.pattern.ExpressionPattern.get_wrappings (mathics/core/pattern.c:9936)
    yield_func(items[0])
  File "mathics/core/pattern.py", line 502, in mathics.core.pattern.ExpressionPattern.match_leaf.yield_wrapping (mathics/core/pattern.c:11242)
    leaf.match(match_yield, item, vars, evaluation, fully=True,
  File "mathics/builtin/patterns.py", line 539, in mathics.builtin.patterns.Pattern_.match (mathics/builtin/patterns.c:9456)
    self.pattern.match(yield_func, expression, new_vars, evaluation)
  File "mathics/builtin/patterns.py", line 701, in mathics.builtin.patterns.Blank.match (mathics/builtin/patterns.c:12459)
    yield_func(vars, None)
  File "mathics/core/pattern.py", line 499, in mathics.core.pattern.ExpressionPattern.match_leaf.match_yield (mathics/core/pattern.c:11104)
    yield_func(new_vars, items_rest)
  File "mathics/core/rules.py", line 74, in mathics.core.rules.BaseRule.apply.yield_match (mathics/core/rules.c:2129)
    new_expression = self.do_replace(vars, options, evaluation)
  File "mathics/core/rules.py", line 165, in mathics.core.rules.BuiltinRule.do_replace (mathics/core/rules.c:4029)
    return self.function(evaluation=evaluation, **vars_noctx)
  File "/home/angus/prog/Mathics/mathics/builtin/lists.py", line 777, in apply
    return expr.leaves[0]
IndexError: list index out of range
```

File: mathics/builtin/lists.py [hunk 2]
```diff
@@ -2012,3 +2012,113 @@ def get_call():
...
+     = 45
+
+    Total over rows instead of columns
+    >> Total[{{1, 2, 3}, {4, 5, 6}, {7, 8 ,9}}, {2}]
+     = {6, 15, 24}
+    """
+    rules = {
+        'Total[list_List]': 'Fold[Plus, list]',
+        'Total[list_List, n_]': 'Fold[Plus, Flatten[list, n]]'
+    }
```
Comment (sn6uv) [inline mathics/builtin/lists.py]: This one also fails on the empty list.
```
In[2]:= Total[{}]
Traceback (most recent call last):
  File "mathics/main.py", line 324, in <module>
    main()
  File "mathics/main.py", line 315, in main
    shell.evaluate(total_input)
  File "mathics/main.py", line 120, in evaluate
    out_callback=out_callback)
  File "/home/angus/prog/Mathics/mathics/core/evaluation.py", line 207, in __init__
    result = evaluate()
  File "/home/angus/prog/Mathics/mathics/core/evaluation.py", line 192, in evaluate
    result = query.evaluate(self)
  File "mathics/core/expression.py", line 776, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:22266)
    result = result.evaluate(evaluation)
  File "mathics/core/expression.py", line 776, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:22266)
    result = result.evaluate(evaluation)
  File "mathics/core/expression.py", line 726, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:21079)
    leaves[index] = leaves[index].evaluate(evaluation)
  File "mathics/core/expression.py", line 773, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:22179)
    result = rule.apply(new, evaluation, fully=False)
  File "mathics/core/rules.py", line 115, in mathics.core.rules.BaseRule.apply (mathics/core/rules.c:2677)
    self.pattern.match(
  File "mathics/core/pattern.py", line 224, in mathics.core.pattern.ExpressionPattern.match (mathics/core/pattern.c:6424)
    self.head.match(
  File "mathics/core/pattern.py", line 153, in mathics.core.pattern.AtomPattern.match (mathics/core/pattern.c:4758)
    yield_func(vars, None)
  File "mathics/core/pattern.py", line 216, in mathics.core.pattern.ExpressionPattern.match.yield_head (mathics/core/pattern.c:6024)
    self.get_pre_choices(
  File "mathics/core/pattern.py", line 341, in mathics.core.pattern.ExpressionPattern.get_pre_choices (mathics/core/pattern.c:8974)
    yield_func(vars)
  File "mathics/core/pattern.py", line 201, in mathics.core.pattern.ExpressionPattern.match.yield_choice (mathics/core/pattern.c:5833)
    self.match_leaf(
  File "mathics/core/pattern.py", line 506, in mathics.core.pattern.ExpressionPattern.match_leaf (mathics/core/pattern.c:12726)
    self.get_wrappings(
  File "mathics/core/pattern.py", line 362, in mathics.core.pattern.ExpressionPattern.get_wrappings (mathics/core/pattern.c:9936)
    yield_func(items[0])
  File "mathics/core/pattern.py", line 502, in mathics.core.pattern.ExpressionPattern.match_leaf.yield_wrapping (mathics/core/pattern.c:11242)
    leaf.match(match_yield, item, vars, evaluation, fully=True,
  File "mathics/builtin/patterns.py", line 539, in mathics.builtin.patterns.Pattern_.match (mathics/builtin/patterns.c:9456)
    self.pattern.match(yield_func, expression, new_vars, evaluation)
  File "mathics/builtin/patterns.py", line 701, in mathics.builtin.patterns.Blank.match (mathics/builtin/patterns.c:12459)
    yield_func(vars, None)
  File "mathics/core/pattern.py", line 499, in mathics.core.pattern.ExpressionPattern.match_leaf.match_yield (mathics/core/pattern.c:11104)
    yield_func(new_vars, items_rest)
  File "mathics/core/rules.py", line 74, in mathics.core.rules.BaseRule.apply.yield_match (mathics/core/rules.c:2129)
    new_expression = self.do_replace(vars, options, evaluation)
  File "mathics/core/rules.py", line 165, in mathics.core.rules.BuiltinRule.do_replace (mathics/core/rules.c:4029)
    return self.function(evaluation=evaluation, **vars_noctx)
  File "/home/angus/prog/Mathics/mathics/builtin/lists.py", line 777, in apply
    return expr.leaves[0]
IndexError: list index out of range
```

File: mathics/builtin/lists.py [hunk 3]
```diff
@@ -2012,3 +2012,113 @@ def get_call():
...
+    >> FoldList[f, x, {1, 2, 3}]
+     = {x, f[x, 1], f[f[x, 1], 2], f[f[f[x, 1], 2], 3]}
+    >> FoldList[Times, {1, 2, 3}]
+     = {1, 2, 6}
+    """
+
+    rules = {
+        'FoldList[exp_, x_, list_List]': 'Prepend[Table[Fold[exp, x, Take[list, i]], {i, 1, Length[list]}], x]',
+        'FoldList[exp_, list_List]': 'FoldList[exp, First[list], Rest[list]]',
+    }
```
Comment (sn6uv) [inline mathics/builtin/lists.py]: This one fails on the empty list
```
In[7]:= FoldList[f, {}]
Traceback (most recent call last):
  File "mathics/main.py", line 324, in <module>
    main()
  File "mathics/main.py", line 315, in main
    shell.evaluate(total_input)
  File "mathics/main.py", line 120, in evaluate
    out_callback=out_callback)
  File "/home/angus/prog/Mathics/mathics/core/evaluation.py", line 207, in __init__
    result = evaluate()
  File "/home/angus/prog/Mathics/mathics/core/evaluation.py", line 192, in evaluate
    result = query.evaluate(self)
  File "mathics/core/expression.py", line 776, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:22266)
    result = result.evaluate(evaluation)
  File "mathics/core/expression.py", line 726, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:21079)
    leaves[index] = leaves[index].evaluate(evaluation)
  File "mathics/core/expression.py", line 773, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:22179)
    result = rule.apply(new, evaluation, fully=False)
  File "mathics/core/rules.py", line 115, in mathics.core.rules.BaseRule.apply (mathics/core/rules.c:2677)
    self.pattern.match(
  File "mathics/core/pattern.py", line 224, in mathics.core.pattern.ExpressionPattern.match (mathics/core/pattern.c:6424)
    self.head.match(
  File "mathics/core/pattern.py", line 153, in mathics.core.pattern.AtomPattern.match (mathics/core/pattern.c:4758)
    yield_func(vars, None)
  File "mathics/core/pattern.py", line 216, in mathics.core.pattern.ExpressionPattern.match.yield_head (mathics/core/pattern.c:6024)
    self.get_pre_choices(
  File "mathics/core/pattern.py", line 341, in mathics.core.pattern.ExpressionPattern.get_pre_choices (mathics/core/pattern.c:8974)
    yield_func(vars)
  File "mathics/core/pattern.py", line 201, in mathics.core.pattern.ExpressionPattern.match.yield_choice (mathics/core/pattern.c:5833)
    self.match_leaf(
  File "mathics/core/pattern.py", line 506, in mathics.core.pattern.ExpressionPattern.match_leaf (mathics/core/pattern.c:12726)
    self.get_wrappings(
  File "mathics/core/pattern.py", line 362, in mathics.core.pattern.ExpressionPattern.get_wrappings (mathics/core/pattern.c:9936)
    yield_func(items[0])
  File "mathics/core/pattern.py", line 502, in mathics.core.pattern.ExpressionPattern.match_leaf.yield_wrapping (mathics/core/pattern.c:11242)
    leaf.match(match_yield, item, vars, evaluation, fully=True,
  File "mathics/builtin/patterns.py", line 539, in mathics.builtin.patterns.Pattern_.match (mathics/builtin/patterns.c:9456)
    self.pattern.match(yield_func, expression, new_vars, evaluation)
  File "mathics/builtin/patterns.py", line 701, in mathics.builtin.patterns.Blank.match (mathics/builtin/patterns.c:12459)
    yield_func(vars, None)
  File "mathics/core/pattern.py", line 499, in mathics.core.pattern.ExpressionPattern.match_leaf.match_yield (mathics/core/pattern.c:11104)
    yield_func(new_vars, items_rest)
  File "mathics/core/rules.py", line 74, in mathics.core.rules.BaseRule.apply.yield_match (mathics/core/rules.c:2129)
    new_expression = self.do_replace(vars, options, evaluation)
  File "mathics/core/rules.py", line 165, in mathics.core.rules.BuiltinRule.do_replace (mathics/core/rules.c:4029)
    return self.function(evaluation=evaluation, **vars_noctx)
  File "/home/angus/prog/Mathics/mathics/builtin/lists.py", line 777, in apply
    return expr.leaves[0]
IndexError: list index out of range
```

File: mathics/builtin/lists.py [hunk 4]
```diff
@@ -2012,3 +2012,113 @@ def get_call():
...
+     = 8
+    >> Fold[f, 5, {1, 2, 3}]
+     = f[f[f[5, 1], 2], 3]
+    """
+
+    attributes = ('NHoldRest',)
+
+    rules = {
+        'Fold[exp_, x_, list_List]': 'Module[{},res=x; Do[res = exp[res, list[[i]]], {i, 1, Length[list]}]; res]',
+        'Fold[exp_, list_List]': u'Fold[exp, First[list], Rest[list]]',
```
Comment (sn6uv) [inline mathics/builtin/lists.py]: - Empty list e.g. `Fold[f, {}]`
- Last argument need not be `List`.

**Comment by sn6uv** [comment]
I've pointed out a couple of issues, but I've realised the empty list issue is not a result of this PR

```
In[1]:= Rest[{}]
Out[1]= {}

In[2]:= First[{}]
Traceback (most recent call last):
  File "mathics/main.py", line 324, in <module>
    main()
  File "mathics/main.py", line 315, in main
    shell.evaluate(total_input)
  File "mathics/main.py", line 120, in evaluate
    out_callback=out_callback)
  File "/home/angus/prog/Mathics/mathics/core/evaluation.py", line 207, in __init__
    result = evaluate()
  File "/home/angus/prog/Mathics/mathics/core/evaluation.py", line 192, in evaluate
    result = query.evaluate(self)
  File "mathics/core/expression.py", line 773, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:22179)
    result = rule.apply(new, evaluation, fully=False)
  File "mathics/core/rules.py", line 115, in mathics.core.rules.BaseRule.apply (mathics/core/rules.c:2677)
    self.pattern.match(
  File "mathics/core/pattern.py", line 224, in mathics.core.pattern.ExpressionPattern.match (mathics/core/pattern.c:6424)
    self.head.match(
  File "mathics/core/pattern.py", line 153, in mathics.core.pattern.AtomPattern.match (mathics/core/pattern.c:4758)
    yield_func(vars, None)
  File "mathics/core/pattern.py", line 216, in mathics.core.pattern.ExpressionPattern.match.yield_head (mathics/core/pattern.c:6024)
    self.get_pre_choices(
  File "mathics/core/pattern.py", line 341, in mathics.core.pattern.ExpressionPattern.get_pre_choices (mathics/core/pattern.c:8974)
    yield_func(vars)
  File "mathics/core/pattern.py", line 201, in mathics.core.pattern.ExpressionPattern.match.yield_choice (mathics/core/pattern.c:5833)
    self.match_leaf(
  File "mathics/core/pattern.py", line 506, in mathics.core.pattern.ExpressionPattern.match_leaf (mathics/core/pattern.c:12726)
    self.get_wrappings(
  File "mathics/core/pattern.py", line 362, in mathics.core.pattern.ExpressionPattern.get_wrappings (mathics/core/pattern.c:9936)
    yield_func(items[0])
  File "mathics/core/pattern.py", line 502, in mathics.core.pattern.ExpressionPattern.match_leaf.yield_wrapping (mathics/core/pattern.c:11242)
    leaf.match(match_yield, item, vars, evaluation, fully=True,
  File "mathics/builtin/patterns.py", line 539, in mathics.builtin.patterns.Pattern_.match (mathics/builtin/patterns.c:9456)
    self.pattern.match(yield_func, expression, new_vars, evaluation)
  File "mathics/builtin/patterns.py", line 701, in mathics.builtin.patterns.Blank.match (mathics/builtin/patterns.c:12459)
    yield_func(vars, None)
  File "mathics/core/pattern.py", line 499, in mathics.core.pattern.ExpressionPattern.match_leaf.match_yield (mathics/core/pattern.c:11104)
    yield_func(new_vars, items_rest)
  File "mathics/core/rules.py", line 74, in mathics.core.rules.BaseRule.apply.yield_match (mathics/core/rules.c:2129)
    new_expression = self.do_replace(vars, options, evaluation)
  File "mathics/core/rules.py", line 165, in mathics.core.rules.BuiltinRule.do_replace (mathics/core/rules.c:4029)
    return self.function(evaluation=evaluation, **vars_noctx)
  File "/home/angus/prog/Mathics/mathics/builtin/lists.py", line 777, in apply
    return expr.leaves[0]
IndexError: list index out of range
```

**Comment by wolfv** [comment]
I have solved the Total and Accumulate: 

        'Accumulate[expr_]': 'Head[expr] @@ FoldList[Plus, Level[expr, 1]]'
        
        'Total[expr_]': 'Fold[Plus, Level[expr, 1]]',
        'Total[expr_, n_]': 'Fold[Plus, Flatten[Level[expr, 1], n]]'

And I am adding a If to the other expressions to solve the empty lists issue. 

In Mathematica First[{}] evaluates to an error as well (it's not crashing, though :)

**Comment by sn6uv** [comment]
You might be able to solve both issues at once with
```
`Accumulate[list_]': 'Rest[FoldList[Plus, 0, list]]'
```
and
```
'Total[expr_]': 'Apply[Plus, expr]'
```

**Comment by sn6uv** [comment]
Really great work by the way. I'm excited to get these features added into Mathics.

**Comment by wolfv** [comment]
Thanks! 

I have, unfortunately, noticed another issue with Flatten:
```
Flatten[p[1, p[2], 3]]
 = p[1, 2, 3]
```

However, in Mathics it currently evaluates to `p[1, p[2], 3]` (nothing changes).

I don't know how easy or hard that is to fix, but if fixed, Total would work as expected :)

**Comment by wolfv** [comment]
Btw. I find the meta-programming capabilities really impressive. It's quite fun to make these functions this way, hehe

**Comment by wolfv** [comment]
Btw. I have one issue with the evaluation of `Fold[f, {}]`. 

What I would like to return is `Defer[Fold[f, {}]]`, ie. the not evaluated equivalent. However, Defer doesn't seem to be implemented currently.

Just wanted to ask if there is another way to return the unevaluated function, or what you'd suggest to do?

**Comment by sn6uv** [comment]
I'll fix that flatten issue right away.

**Comment by wolfv** [comment]
Btw: I have a bunch of other functions implemented here: https://github.com/mathics/Mathics/compare/master...wolfv:more-linalg

The most important ones are (changed) Floor and Ceiling based on sympy. 

Then Cross, and Norm are quite cool. 

I also changed the way the conversion to and from mathics with matrices works. 

I.e. this Cross[{1,2,3}, {4,5,6}] results in {{-3}, {6}, {-3}} instead of {-3, 6, -3}.

Unfortunately it's exam time for me and I should cut down on my computer usage, so I won't have a lot of time to do work on that stuff for the next ~2 month.

**Comment by sn6uv** [comment]
I've opened #269 to fix that Flatten bug. I'll merge once the tests pass (usually ~15min).

**Comment by sn6uv** [comment]
There are 2 ways to return an unevaluated expression:
- return `None` from an `apply_` method.
- match no patterns (includes `apply_` methods and rules).

**Comment by wolfv** [comment]
Cool, that worked. Code that get's shorter is something I like a lot.

**Comment by sn6uv** [comment]
I'll look over this a bit more but I'm getting to the point where I'm happy to merge this.

**Pull request merged**